src/shutdown.rs

```diff
@@ -36,13 +43,15 @@ pub struct ShutdownSignal {
 }
 
 impl Future for ShutdownSignal {
-    type Item = ShutdownSignalToken;
-    type Error = ();
-    fn poll(&mut self) -> Result<Async<Self::Item>, Self::Error> {
-        match self.begin_shutdown.poll() {
-            Ok(Async::Ready(_)) => Ok(Async::Ready(self.shutdown_complete.take().unwrap())),
-            Ok(Async::NotReady) => Ok(Async::NotReady),
-            Err(_) => Err(()),
+    type Output = ShutdownSignalToken;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let closed = ready!(this.begin_shutdown.poll(cx));
+        if closed {
+            Poll::Ready(this.shutdown_complete.take().unwrap())
+        } else {
+            Poll::Pending
```

Comment:
`Tripwire` allow call `poll` after being resolved, for example:
```rust
use stream_cancel::Tripwire;

#[tokio::main]
async fn main() {
    let (trigger, mut tripwire) = Tripwire::new();
    println!("{:?}", futures::poll!(&mut tripwire));
    trigger.cancel();
    println!("{:?}", futures::poll!(&mut tripwire));
    println!("{:?}", futures::poll!(&mut tripwire));
}
```
will print:
```text
Pending
Ready(true)
Ready(true)
```

I think we should return `Poll::Pending` because `ShutdownSignal` shouldn't be resolved if tripwire was disabled. Also, if we return `Poll::Pending` on `Poll::Ready` from `begin_shutdown` wake will never be called on context.

Reply:
> I think we should return Poll::Pending because ShutdownSignal shouldn't be resolved if tripwire was disabled. 

That's fine, but

> Also, if we return Poll::Pending on Poll::Ready from begin_shutdown wake will never be called on context?

can't be guaranteed because it is allowed by `Future` contract, and we shouldn't depend on specific implementation/behavior of `Tripwire`. This would result in a sneaky bug that can be introduced with subsequent updates to `stream-cancel` or other crates. 

So I think a few code lines extra is worthy of avoiding that.

Reply:
Updated. Please check again :slightly_smiling_face: